zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1217 @@
+        );
+        assert_eq!(
+            checkpoint_verifier.target_checkpoint_height(),
+            WaitingForBlocks
+        );
+        assert_eq!(checkpoint_verifier.max_checkpoint_height(), BlockHeight(0));
 
         /// Make sure the verifier service is ready
         let ready_verifier_service = checkpoint_verifier
             .ready_and()
             .await
             .map_err(|e| eyre!(e))?;
-        /// Try to verify block 415000, and expect failure
+        /// Set up the future for block 415000
+        let verify_future = timeout(
+            Duration::from_secs(VERIFY_TIMEOUT_SECONDS),
+            ready_verifier_service.call(block415000.clone()),
+        );
+        /// Wait for the response for block 415000, and expect failure
         // TODO(teor || jlusby): check error kind
-        ready_verifier_service
-            .call(block415000.clone())
+        let _ = verify_future
             .await
-            .unwrap_err();
+            .expect("timeout should not happen")
+            .expect_err("bad block hash should fail");
 
-        /// Make sure the state service is ready (1/2)
-        let ready_state_service = state_service.ready_and().await.map_err(|e| eyre!(e))?;
-        /// Make sure neither block is in the state: expect GetBlock 415000 to fail.
-        // TODO(teor || jlusby): check error kind
-        ready_state_service
-            .call(zebra_state::Request::GetBlock {
-                hash: block415000.as_ref().into(),
-            })
-            .await
-            .unwrap_err();
-
-        /// Make sure the state service is ready (2/2)
-        let ready_state_service = state_service.ready_and().await.map_err(|e| eyre!(e))?;
-        /// Make sure neither block is in the state: expect GetBlock 0 to fail.
-        // TODO(teor || jlusby): check error kind
-        ready_state_service
-            .call(zebra_state::Request::GetBlock {
-                hash: block0.as_ref().into(),
-            })
-            .await
-            .unwrap_err();
+        assert_eq!(
+            checkpoint_verifier.previous_checkpoint_height(),
+            BeforeGenesis
+        );
+        assert_eq!(
+            checkpoint_verifier.target_checkpoint_height(),
+            WaitingForBlocks
+        );
+        assert_eq!(checkpoint_verifier.max_checkpoint_height(), BlockHeight(0));
 
         Ok(())
     }
 
     #[tokio::test]
     #[spandoc::spandoc]
-    async fn checkpoint_wrong_hash_fail() -> Result<(), Report> {
-        install_tracing();
+    async fn wrong_checkpoint_hash_fail() -> Result<(), Report> {
+        zebra_test::init();
 
-        let block0 =
+        let good_block0 =
             Arc::<Block>::zcash_deserialize(&zebra_test::vectors::BLOCK_MAINNET_GENESIS_BYTES[..])?;
+        let good_block0_hash: BlockHeaderHash = good_block0.as_ref().into();
+        let mut bad_block0 =
+            Block::zcash_deserialize(&zebra_test::vectors::BLOCK_MAINNET_GENESIS_BYTES[..])?;
```

Comment:
TIOLI: you can just clone the `Arc` here and then use `Arc::make_mut` to do a copy on write while mutating the version

https://doc.rust-lang.org/stable/std/sync/struct.Arc.html#method.make_mut

Reply:
Fixed in c7d4b877f063c9fb28b03d9bdc42b101dc0d1aff.

It is surprisingly easy to accidentally modify the original block.